[sayglenn] iP

sayglenn opened the pull request:
# Ginger
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

_Need some help with your tasks?_ Ginger is here to help you remember your tasks and organise them better!

## Features
Ginger has **many** features available to users, which include:
- Adding / Deleting Tasks
- Listing Tasks
- Storing Tasks in Local Database
- Marking / Unmarking of Tasks

## How to Install
1. Download the ginger.jar file from [here.](https://github.com/sayglenn/ip/releases/tag/A-Jar)
2. In your terminal, create a folder using `mkdir` and enter it using `cd`.
3. Place the ginger.jar file into the folder.
4. To start Ginger, run:
```
java -jar ginger.jar
```
5. Feel free to explore the chat bot! 😄

## Upcoming Plans
Ginger intends to add more features to complement your tasks.
- [X] Managing Tasks
- [ ] Managing Deadlines (coming soon!)
- [ ] Reminders (coming soon!)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

sayglenn committed:
Remove Duke bot and add Ginger bot

sayglenn committed:
Add scanner to take in user input

sayglenn committed:
Enable repeated input until bye is entered

sayglenn committed:
Create task storing and retrieval logic

sayglenn committed:
Integrate task list with Ginger

sayglenn committed:
Add task class

sayglenn committed:
Add input processor to manage mark and unmark

sayglenn committed:
Add logic to handle out-of-bounds task marking/unmarking

sayglenn committed:
Amend status change for task method

sayglenn committed:
Add new task types

sayglenn committed:
Complete integration of different task into Ginger

sayglenn committed:
Add automated text UI testing

sayglenn committed:
Add command enum for improved flow

sayglenn committed:
Add exception handling for incorrect arguments and commands

sayglenn committed:
Reconfigure exception flow and increase automated text ui testing

sayglenn committed:
Add delete function and amend automated text UI testing

sayglenn committed:
Change Task.java to abstract and add Javadocs

sayglenn committed:
Add function to save to local database

sayglenn committed:
Merge branch 'branch-Level-7'

sayglenn committed:
Add support for date and time for clearer task info

sayglenn committed:
Merge branch 'branch-Level-8'

sayglenn committed:
Update help message to reflect new features

sayglenn committed:
Abstract database reading to separate class

sayglenn committed:
Add TaskHandler class to abstract handling of task list

sayglenn committed:
Add task addition to TaskHandler

sayglenn committed:
Add command classes and modify task handlers and storage classes

sayglenn committed:
Add Ui class to abstract user interface handling

sayglenn committed:
Add InputParser class to handle input and cleaned up other code

sayglenn committed:
Merge branch 'branch-A-MoreOOP'

sayglenn committed:
Repackage classes for increased readability and ease of navigation

sayglenn committed:
Amend help message and corrected text-ui-test expected results

sayglenn committed:
Merge branch 'add-gradle-support'

sayglenn committed:
Modify build.gradle to reflect the correct main class

sayglenn committed:
Add equals method for commands for comparison

sayglenn committed:
Add tests for input parsing and task handling

sayglenn committed:
Add javadocs for classes and methods

sayglenn committed:
Amend parameters and methods to align with boolean coding standards

sayglenn committed:
Add find function to Ginger

sayglenn committed:
Merge branch 'branch-Level-9'

sayglenn committed:
Merge branch 'branch-A-JavaDoc'

sayglenn committed:
Merge branch 'branch-A-CodingStandard'

sayglenn committed:
Add test for new find command

sayglenn committed:
Add checkstyle to Ginger for better code quality

sayglenn committed:
Correct code according to checkstyle requirements

IzN432 commented on src/test/java/ginger/task/TaskHandlerTest.java:
```diff
@@ -0,0 +1,23 @@
+public class TaskHandlerTest {
+
+    @Test
+    public void TaskHandlerTest_TaskAddsCorrectly() {
+        TaskHandler taskHandler = new TaskHandler();
+
+        // Clear any potential pre-existing tasks for proper testing
+        for (int i = 0; i < taskHandler.taskCount(); i++) {
+            taskHandler.deleteTask(0);
+        }
+
+        Task t = new ToDo("buy groceries");
+        taskHandler.addTask(t);
+
+        assertEquals(taskHandler.getTask(0), t);
+    }
```
Perhaps the method name should be in camel-case?

IzN432 commented on src/test/java/ginger/parser/InputParserTest.java:
```diff
@@ -0,0 +1,77 @@
+package ginger.parser;
+
+import ginger.command.*;
```
Would it be clearer if you imported each class individually rather than used a wildcard import statement?

IzN432 commented on src/main/java/ginger/ui/Ui.java:
```diff
@@ -0,0 +1,24 @@
+ * Represents the UI of the bot, which will display replies to the user input in the terminal.
+ */
+public class Ui {
+    private final String HORIZONTAL_LINE = "____________________________________________________________";
+    private final String BOT_NAME = "Ginger";
```
Perhaps these fields should be static?

IzN432 commented on src/main/java/ginger/task/TaskHandler.java:
```diff
@@ -0,0 +1,149 @@
+     */
+    public Task addToDo(String title) {
+        ToDo newToDo = new ToDo(title);
+        this.addTask(newToDo);
```
Perhaps you could leave out ```this.```?

IzN432 commented on src/main/java/ginger/task/TaskHandler.java:
```diff
@@ -0,0 +1,149 @@
+     * @param taskList The task list to be converted.
+     * @return The task list in a string format.
+     */
+    public String tasklistToString(List<Task> taskList) {
```
Maybe the method name could be ```taskListToString``` with the capitalized List instead?

IzN432 commented on src/main/java/ginger/task/Event.java:
```diff
@@ -0,0 +1,50 @@
+    public String toString() {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd MMM yyyy, HH:mm");
+        return String.format("[E]%s (%s - %s)", super.toString(),
+                this.start.format(formatter), this.end.format(formatter));
```
In general, I like to leave out unnecessary ```this.``` but it is not exactly in the coding standards so feel free to ignore

IzN432 commented on src/test/java/ginger/parser/InputParserTest.java:
```diff
@@ -0,0 +1,77 @@
+package ginger.parser;
+
+import ginger.command.*;
+import ginger.exception.IllegalGingerArgumentException;
+import ginger.exception.IllegalGingerCommandException;
+import org.junit.jupiter.api.Test;
+import java.time.LocalDateTime;
```
Perhaps standard library imports should go above and be separated by a line?

IzN432 reviewed (commented):
LGTM. Overall the code was easy to follow and follows the code quality guidelines except for some tiny nits. Well done!

1st2GetThisName commented on src/main/java/ginger/command/Command.java:
```diff
@@ -0,0 +1,16 @@
+public abstract class Command {
+    public abstract void execute(TaskHandler taskHandler, Ui ui) throws IllegalGingerArgumentException;
+
+    public boolean isExit() {
+        return false;
+    }
```
Does isExit represent whether a command causes the application to exit?

1st2GetThisName commented on src/main/java/ginger/parser/InputParser.java:
```diff
@@ -0,0 +1,193 @@
+ */
+public class InputParser {
+    enum CommandList {
+        BYE, LIST, MARK, UNMARK, TODO, DEADLINE, EVENT, HELP, DELETE, FIND;
```
Good job remembering to capitalize your enums!

1st2GetThisName reviewed (commented):
LGTM, not much to comment on, and no news is good news!

sayglenn committed:
Build GUI for user input

sayglenn committed:
Merge branch 'branch-Level-10'

sayglenn committed:
Amend GUI to greet user upon initialising

sayglenn committed:
Add Launcher class and amend test to correct behaviour

sayglenn committed:
Add assertions to code to validate functionality

Certain commands, such as addToDo, should result in specific behaviour
like resulting in the task list being non-empty.

Using a form of checking the current state of the program will help
ensure such behaviour is followed and no unexpected behaviour will
occur.

Let's use assertions to ensure the behaviour of the code can be
predicted.

By using assertions, we can ensure the code behaves as originally
intended. Some examples include:
* Functions addToDo, addDeadline, addEvent should result in a task list
  that is non-empty, we assert that such behaviour is true.
* Initiation of the TaskHandler class should create a List<Task> object
  and populate the taskList field to prevent NullPointerException.
* Using getCommand in InputParser should always return a CommandList
  object as an exception would be thrown on invalid commands.

Through assertions, we can maintain the expected behaviour of code and
ensure that it functions as intended.

sayglenn committed:
Fix checkstyle issue regarding imports and constants

sayglenn committed:
Refactor code to align with code quality standards

Code in the repository can always be looked for improvement. Aligning to
standards such as LoC for method lengths, SLAP and not deeply nesting is
important to ensure high quality of code.

By properly following good coding quality practices, it ensures
readability of code and maintains standards for future features.

Let's refactor the parse method in the InputParser class to align to
these coding standards by reducing the LoC.

In order to reduce the LoC in the method:
* Common checks such as checking the number of arguments in input parts
  will be extracted to its individual method, validateArguments.
* For each case in the switch, extract the check of number of arguments
  and returning of the Command to its respective method.

Through extracting commonly used code, we heavily reduce the number of
lines required in the original parse method and align with coding
quality standards.

sayglenn committed:
Add streams to simplify display of task list

sayglenn committed:
Merge pull request #1 from sayglenn/branch-A-Assertions

Add assertions to ensure expected behaviour occurs

sayglenn committed:
Merge pull request #2 from sayglenn/branch-A-CodeQuality

Refactor code for increased code quality

sayglenn committed:
Merge remote-tracking branch 'origin/master' into branch-A-Streams

sayglenn committed:
Merge pull request #3 from sayglenn/branch-A-Streams

Add stream to simplify method logic

sayglenn committed:
Integrate CI into Ginger

sayglenn committed:
Add feature to tag tasks

sayglenn committed:
Merge branch 'branch-BCD-Extension'

sayglenn committed:
Customise GUI and add error css for invalid commands

sayglenn committed:
Add more Ginger themed replies for better personality

sayglenn committed:
Change commands to save tasks upon every execute

sayglenn committed:
Add Ui image and update readme

sayglenn committed:
Fix checkstyle errors

sayglenn committed:
Update readme with latest release

sayglenn committed:
Fix parser to disallow empty fields and amend task display

sayglenn committed:
Fix parsing in case of empty title with event or deadline

sayglenn committed:
Add tests and more line breaks to help message

sayglenn committed:
Amend method name to align with camel case convention
